File: a2/383A2/exer2.erl
-module(exer2).
-export([divisors/1, primes/1, join/2, pythagorean/1]).

divisors(N) ->
  [I || I <- lists:seq(2, N div 2), N rem I =:= 0].

primes(N) ->
  [I || I <- lists:seq(2, N), divisors(I) =:= []].

join(_, []) -> [];
join(_, [X]) -> X;
join(A, [X|XS]) -> X ++ A ++ join(A, XS).

pythagorean(N) ->
  [{A,B,C} || C <- lists:seq(1, N), A <- lists:seq(1, C), B <- lists:seq(A, C), (A * A)+(B * B)==(C * C)].

File: exer2/exer2.hs
divisors :: Int -> [Int]
divisors n = [i | i <- [2..(n `div` 2)], n `mod` i == 0]
primes :: Int -> [Int]
primes n = [i | i <- [2..n], divisors i == []]

pythagorean :: Int -> [(Int, Int, Int)]
pythagorean n = [(a,b,c) | c <- [1..n], a <- [1..c], b <- [a..c], (a^2)+(b^2)==(c^2)]

join :: [Char] -> [[Char]] -> [Char]
join _ ([]) = []
join _ (x:[]) = x
join a (x:xs) = x ++ a ++ join a xs

fact' :: Int -> Int
fact' n = foldl (*) 1 [1..n]

hailstone n
    | even n        = div n 2
    | otherwise     = 3 * n + 1

hailLen n = hailTail 0 n
  where
    hailTail a 1 = a
    hailTail a x = hailTail (a+1) (hailstone x)